--- hmp-commands.hx	2018-12-11 11:44:34.000000000 -0600
+++ hmp-commands.hx.patched	2019-04-04 16:41:48.000000000 -0500
@@ -1897,6 +1897,17 @@
         .flags      = "p",
     },
 
+#if defined(CONFIG_PROCESSOR_TRACE) && defined(TARGET_X86_64)
+    {
+        .name       = "pt",
+        .args_type  = "item:s?",
+        .params     = "[subcommand]",
+        .help       = "show various option to configure Intel Processor Tracing",
+        .cmd        = hmp_info_help,
+        .sub_table = pt_cmds,
+    },
+#endif
+
 STEXI
 @end table
 ETEXI
