--- Makefile.target	2018-12-11 11:44:34.000000000 -0600
+++ Makefile.target.patched	2019-04-04 17:23:10.000000000 -0500
@@ -145,7 +145,8 @@
 obj-y += dump.o
 obj-$(TARGET_X86_64) += win_dump.o
 obj-y += migration/ram.o
-LIBS := $(libs_softmmu) $(LIBS)
+obj-$(CONFIG_PROCESSOR_TRACE) += pt/memory_access.o pt/decoder.o pt/tnt_cache.o pt/disassembler.o pt/logger.o pt/hypercall.o pt/filter.o pt.o pt/interface.o
+LIBS := $(libs_softmmu) $(LIBS) -lcapstone
 
 # Hardware support
 ifeq ($(TARGET_NAME), sparc64)
@@ -154,7 +155,7 @@
 obj-y += hw/$(TARGET_BASE_ARCH)/
 endif
 
-GENERATED_FILES += hmp-commands.h hmp-commands-info.h
+GENERATED_FILES += hmp-commands.h hmp-commands-info.h hmp-commands-pt.h
 
 endif # CONFIG_SOFTMMU
 
@@ -208,6 +209,9 @@
 hmp-commands.h: $(SRC_PATH)/hmp-commands.hx $(SRC_PATH)/scripts/hxtool
 	$(call quiet-command,sh $(SRC_PATH)/scripts/hxtool -h < $< > $@,"GEN","$(TARGET_DIR)$@")
 
+hmp-commands-pt.h: $(SRC_PATH)/hmp-commands-pt.hx
+	$(call quiet-command,sh $(SRC_PATH)/scripts/hxtool -h < $< > $@,"  GEN   $(TARGET_DIR)$@")
+
 hmp-commands-info.h: $(SRC_PATH)/hmp-commands-info.hx $(SRC_PATH)/scripts/hxtool
 	$(call quiet-command,sh $(SRC_PATH)/scripts/hxtool -h < $< > $@,"GEN","$(TARGET_DIR)$@")
 
