--- monitor.c	2018-12-11 11:44:34.000000000 -0600
+++ monitor.c.patched	2019-04-04 16:54:37.000000000 -0500
@@ -276,6 +276,9 @@
 
 static mon_cmd_t mon_cmds[];
 static mon_cmd_t info_cmds[];
+#if defined(CONFIG_PROCESSOR_TRACE)
+static mon_cmd_t pt_cmds[];
+#endif
 
 QmpCommandList qmp_commands, qmp_cap_negotiation_commands;
 
@@ -2616,6 +2619,13 @@
     { NULL, NULL, },
 };
 
+#if defined(CONFIG_PROCESSOR_TRACE)
+static mon_cmd_t pt_cmds[] = {
+#include "hmp-commands-pt.h"
+    { NULL, NULL, },
+};
+#endif
+
 /*******************************************************************/
 
 static const char *pch;
