--- kvm_host.h	2019-03-27 00:18:00.000000000 -0500
+++ kvm_host.h.patched	2019-04-03 17:19:46.000000000 -0500
@@ -1196,6 +1196,11 @@
 	int (*nested_enable_evmcs)(struct kvm_vcpu *vcpu,
 				   uint16_t *vmcs_version);
 	uint16_t (*nested_get_evmcs_version)(struct kvm_vcpu *vcpu);
+	
+#ifdef CONFIG_KVM_VMX_PT
+	int (*setup_trace_fd)(struct kvm_vcpu *vcpu);
+	int (*vmx_pt_enabled)(void);
+#endif
 };
 
 struct kvm_arch_async_pf {
